### PR TITLE
C# changes for hybrid action spaces

### PR DESCRIPTION
**Pull request by dongruoping**
### Proposed change(s)

C# changes for hybrid action spaces

### Types of change(s)

- [ ] Bug fix
- [x] New feature
- [ ] Code refactor
- [ ] Breaking change
- [ ] Documentation update
- [ ] Other (please describe)

### Checklist
- [ ] Added tests that prove my fix is effective or that my feature works
- [ ] Updated the [changelog](https://github.com/Unity-Technologies/ml-agents/blob/master/com.unity.ml-agents/CHANGELOG.md) (if applicable)
- [ ] Updated the [documentation](https://github.com/Unity-Technologies/ml-agents/tree/master/docs) (if applicable)
- [ ] Updated the [migration guide](https://github.com/Unity-Technologies/ml-agents/blob/master/docs/Migrating.md) (if applicable)

### CODE REVIEW

File: com.unity.ml-agents/Runtime/Communicator/GrpcExtensions.cs
Comment (chriselion) [inline com.unity.ml-agents/Runtime/Communicator/GrpcExtensions.cs:106]: nit: I don't think we should deprecate ActionDescriptions (unless we're replacing them with more specific ContinuousActionDescriptions , etc). We don't currently use them, but I think they will be helpful for debugging in the future, especially now that we can have multiple actuators and action types.

**Review by chriselion — APPROVED** [review]
Looks good. I want to do a pass before we merge to master on a few things like the error messages, and whether we want "hybrid" to be the actual user-facing name. But that shouldn't block this for now.

File: com.unity.ml-agents/Runtime/Communicator/GrpcExtensions.cs
Comment (dongruoping) [reply com.unity.ml-agents/Runtime/Communicator/GrpcExtensions.cs:106]: I changed this because we won't use it anywhere other than deprecated cases since it's for VectorAction, and if we ever need a description we should add Continuous(/Discrete)ActionDescriptions

File: com.unity.ml-agents/Runtime/Communicator/UnityRLCapabilities.cs
Comment (vincentpierre) [inline com.unity.ml-agents/Runtime/Communicator/UnityRLCapabilities.cs:17]:
```suggestion
        public UnityRLCapabilities(
            bool baseRlCapabilities = true, 
            bool concatenatedPngObservations = true,
            bool compressedChannelMapping = true, 
            bool hybridActions = true)
```

File: com.unity.ml-agents/Runtime/Inference/BarracudaModelExtensions.cs
Comment (vincentpierre) [inline com.unity.ml-agents/Runtime/Inference/BarracudaModelExtensions.cs:285]: I wonder if this is future proof. There will be some more deprecations in the future. Maybe using a model version or calling this `NoHybridSupport`?

**Review by vincentpierre — APPROVED** [review]

File: com.unity.ml-agents/Runtime/Inference/BarracudaModelExtensions.cs
Comment (dongruoping) [reply com.unity.ml-agents/Runtime/Inference/BarracudaModelExtensions.cs]: Agreed, but don't like the negative term. I changed it to `HasHybridSupport()`.

File: com.unity.ml-agents/Runtime/Communicator/GrpcExtensions.cs
Comment (surfnerd) [inline com.unity.ml-agents/Runtime/Communicator/GrpcExtensions.cs:136]: we support hybrid actions if the TrainerCapabilities is no?

Comment (surfnerd) [reply com.unity.ml-agents/Runtime/Communicator/GrpcExtensions.cs:136]: or is it supposed to be `!= &&`

Comment (chriselion) [reply com.unity.ml-agents/Runtime/Communicator/GrpcExtensions.cs:136]: TrainerCapabilities is null if there's no trainer connected. So in that case, we support whatever code is currently running in the package.

**Pull request merged**
